libnavigation-base/src/main/java/com/mapbox/navigation/base/route/NavigationRoute.kt

```diff
@@ -5,28 +5,192 @@ package com.mapbox.navigation.base.route
+import com.mapbox.api.directions.v5.models.LegStep
 import com.mapbox.api.directions.v5.models.RouteLeg
 import com.mapbox.api.directions.v5.models.RouteOptions
-import java.util.UUID
+import com.mapbox.api.directions.v5.models.StepIntersection
+import com.mapbox.api.directions.v5.models.StepManeuver
+import com.mapbox.geojson.LineString
+import com.mapbox.geojson.Point
+import com.mapbox.navigation.base.internal.NativeRouteParserWrapper
+import com.mapbox.navigation.base.internal.SDKRouteParser
+import com.mapbox.navigation.utils.internal.InternalJobControlFactory
+import com.mapbox.navigator.RouteInterface
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.runBlocking
+import kotlinx.coroutines.suspendCancellableCoroutine
+import java.net.URL
+import kotlin.coroutines.resume
+import kotlin.coroutines.resumeWithException
 
 /**
  * Wraps a route object used across the Navigation SDK features.
- *
  * @param directionsResponse the original response that returned this route object.
  * @param routeIndex the index of the route that this wrapper tracks
  * from the collection of routes returned in the original response.
  * @param routeOptions options used to generate the [directionsResponse]
- * @param id unique ID of this route entity used for matching it against additional metadata.
- * Unique ID doesn't mean that the underlying [directionsRoute] is unique.
- * Multiple [NavigationRoute]s can represent similar or equal [DirectionsRoute].
  */
-class NavigationRoute(
+class NavigationRoute internal constructor(
     val directionsResponse: DirectionsResponse,
     val routeIndex: Int,
     val routeOptions: RouteOptions,
-    val id: String = UUID.randomUUID().toString()
+    internal val nativeRoute: RouteInterface,
 ) {
 
+    companion object {
+        private val jobControl = InternalJobControlFactory.createDefaultScopeJobControl()
+
+        /**
+         * Creates new instances of [NavigationRoute] based on the routes found in the [directionsResponse].
+         *
+         * @param directionsResponse response to be parsed into [NavigationRoute]s
+         * @param routeOptions options used to generate the [directionsResponse]
+         * @param callback to get notified with the result of parsing
+         */
+        @JvmStatic
+        fun create(
+            directionsResponse: DirectionsResponse,
+            routeOptions: RouteOptions,
+            callback: OnNavigationRoutesReadyCallback
+        ): Job {
+            return jobControl.scope.launch {
+                val navigationRoutes = create(
+                    directionsResponse,
+                    directionsResponseJson = directionsResponse.toJson(),
+                    routeOptions,
+                    routeOptionsUrlString = routeOptions.toUrl("").toString()
+                )
+                callback.onNavigationRoutesReady(navigationRoutes)
+            }
+        }
```

Comment:
What if make all the `create` methods blocking with a warning that there is a serialisation under the hood which can take some time? This will simplify `NavigationRoute` and let users will pick a convenient way for him to execute it using background thread. Right now we force them to use hybrid coroutines/callback approach.

Reply:
I don't think we should push threading down to developers. We can expose a synchronous function though if someone does prefer managing that themselves :+1: 

That said, I wasn't sure about exposing the `Job`, especially from a Java consumers perspective. It works just fine for cancellations, but is a little bit clunky for anything else. Open to suggestions!

Reply:
I understand, you want to make dev's life easier. What if instead of mixed coroutines/callback api we provide 2 functions:
1. coroutine friendly `create` suspend function which does serialisation/deserialisation using `Dispatchers.Default`
2. simple blocking `create` for users of other multithreading approaches(RxJava for example) or for java users.

The first can be a wrapper of the second

Reply:
One more argument for blocking create. `DirectionsResponse.fromJson` is blocking already.  When somebody (for example how we do in `RouterWrapper`) creates `NavigationRoute` he already has to manually schedule work to a background thread. Why not to make `create` consistent with what we already do?

Reply:
That's a good reasoning for me - developers would already have to manage the deserialization workload. Let's keep it that way for the first iteration, we can always expand the API surface later if needed.

Reply:
@LukasPaczos, we will be able to add new methods later, but we won't be able to change existing API. The inability of changing API is the reason I find this discussion important now. 

What if we at least get rid of explicit dependency on Kotlin Coroutines and return custom type instead of a `Job`:
```kotlin
interface CancelactionToken {
    fun cancel()
}

internal class JobCancellationToken(private val job: Job): CancelationToken {
    override fun cancel() {
        job.cancel()
    }
}
```

So that the method will look like this:
```kotin
        @JvmStatic
        fun create(
            directionsResponse: DirectionsResponse,
            routeOptions: RouteOptions,
            callback: OnNavigationRoutesReadyCallback
        ): CancelationToken {
            val job = jobControl.scope.launch {
                val navigationRoutes = create(
                    directionsResponse,
                    directionsResponseJson = directionsResponse.toJson(),
                    routeOptions,
                    routeOptionsUrlString = routeOptions.toUrl("").toString()
                )
                callback.onNavigationRoutesReady(navigationRoutes)
            }
            return JobCancelationToken(job)
        }
```

Reply:
Sorry, I didn't notice that the code is changed already 🙂